sound/soc/sof/loader.c

```diff
@@ -131,6 +131,57 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 }
 EXPORT_SYMBOL(snd_sof_fw_parse_ext_data);
 
+static int xman_get_fw_version(struct snd_sof_dev *sdev,
+			       struct xman_elem_header *hdr)
+{
+	struct xman_fw_version *v = container_of(hdr, struct xman_fw_version,
+						 hdr);
+
+	/* make sure ABI version is compatible */
+	dev_info(sdev->dev, "Firmware info: version %d:%d:%d-%s\n",
+		 v->major, v->minor, v->micro, v->tag);
+	dev_info(sdev->dev,
+		 "Firmware: ABI %d:%d:%d Kernel ABI %d:%d:%d\n",
+		 SOF_ABI_VERSION_MAJOR(v->abi_version),
+		 SOF_ABI_VERSION_MINOR(v->abi_version),
+		 SOF_ABI_VERSION_PATCH(v->abi_version),
+		 SOF_ABI_MAJOR, SOF_ABI_MINOR, SOF_ABI_PATCH);
+
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, v->abi_version)) {
+		dev_err(sdev->dev, "error: incompatible FW ABI version\n");
+		return -EINVAL;
+	}
+
+	if (v->abi_version > SOF_ABI_VERSION) {
+		if (!IS_ENABLED(CONFIG_SND_SOC_SOF_STRICT_ABI_CHECKS)) {
+			dev_warn(sdev->dev, "warn: FW ABI is more recent than kernel\n");
+		} else {
+			dev_err(sdev->dev, "error: FW ABI is more recent than kernel\n");
+			return -EINVAL;
+		}
+	}
+
+	if (v->flags & SOF_IPC_INFO_BUILD) {
+		dev_info(sdev->dev,
+			 "Firmware debug build %d on %s-%s - options:\n"
+			 " GDB: %s\n"
+			 " lock debug: %s\n"
+			 " lock vdebug: %s\n",
+			 v->build, v->date, v->time,
+			 (v->flags & SOF_IPC_INFO_GDB) ?
+				"enabled" : "disabled",
+			 (v->flags & SOF_IPC_INFO_LOCKS) ?
+				"enabled" : "disabled",
+			 (v->flags & SOF_IPC_INFO_LOCKSV) ?
+				"enabled" : "disabled");
```

Comment:
that all looks like duplicated code?

Reply:
Yes, but with new input data type. Old code is kept to be backward compatible.

Reply:
can't we try something smarter than copy-paste to deal with just a type difference?

Reply:
changed to use sof_ipc_* struct and old parsing functions